src/crimson/os/seastore/seastore_types.h

```diff
@@ -1025,7 +1026,120 @@ inline extent_len_le_t init_extent_len_le(extent_len_t len) {
+using local_object_id_le_t = ceph_le32;
+
+constexpr int OFFSET_BITS_NULL = std::numeric_limits<int>::max();
+
+using shard_t = int8_t;
+using pool_t = int64_t;
+// Note: this is the reversed version of the object hash
+using crush_hash_t = uint32_t;
+using snap_t = uint64_t;
+using gen_t = uint64_t;
+static_assert(sizeof(shard_t) == sizeof(ghobject_t().shard_id.id));
+static_assert(sizeof(pool_t) == sizeof(ghobject_t().hobj.pool));
+static_assert(sizeof(crush_hash_t) == sizeof(ghobject_t().hobj.get_bitwise_key_u32()));
+static_assert(sizeof(snap_t) == sizeof(ghobject_t().hobj.snap.val));
+static_assert(sizeof(gen_t) == sizeof(ghobject_t().generation));
+
+/**
+ * laddr_t
+ *
+ * laddr_t represents the logical address of a LogicalCachedExtent.
+ *
+ * # Static Layout
+ * laddr_t uses a 128-bit integer to represent the address value internally.
+ * Besides the basic properties inherits from integers, such as strong ordering
+ * and arithmetic operations, some types of laddr_t also includes the properties
+ * derived from user data. The static layout design ensures these properties are
+ * deterministic and predictable, allowing more optimization based on laddr_t.
+ *
+ * # Overview
+ * The layout of laddr_t has three parts:
+ *
+ * [upgrade:1][object_info:79][object_content:48]
+ *
+ * Each pair of square brackets represents a property stored within the laddr_t
+ * interger. Each property contains a name and the count of bits it used.
+ *
+ * When the object info for different objects does not conflict, we obtain a
+ * useful property:
+ * Each object could have a unique laddr prefix within an OSD.
+ *
+ * Base on this property, we could:
+ * 1. Group different snapshots of on object to the same laddr prefix, which
+ *    could speed up cloning process.
+ * 2. Track frequently accessed objects using laddr_t without keeping the full
+ *    object name.
+ * 3. Effectively remove extents of an object via range deletion.
+ *
+ * # Object Info
+ * The defination of this property is:
+ * [pool:12][shard:8][reverse_hash:32][local_object_id:27]
+ *
+ * The pool, shard and reverse_hash come from the information of object;
+ * The local_object_id is just a random number to handle the collision between
+ * objects. When a collision occurs, the local clone ids of conflicting objects
+ * should be different.
+ *
+ * For global meta logical extents, such as RootBlock, CollectionNode and
+ * SeastoreNodeExtent, these fields should be set to zero, and other objects
+ * data should never use this prefix.
+ *
+ * # Object Content
+ * TODO: This property might change at runtime.
+ *
+ * The defination of this property is:
+ * [is_metadata:1][object_data:47]
+ *
+ * The `is_metadata` field is fixed and always located at the most significant
+ * bit of the object content, so it can't be changed at runtime.
+ *
+ * ## is_metadata true
+ * The remaining bits represent the address of OMap*Node. Take a random value
+ * as address when allocating a new omap extent.
```

Comment:
The extent size might be exceed 4K. I think we should be careful at the boundary. For example, we need to validate whether the end address violates the expected laddr range, right?

Reply:
Yes, related logic is handled by allocating extents in LBAManager.